lib/RateLimiterEtcd.js

```diff
@@ -0,0 +1,113 @@
+const RateLimiterStoreAbstract = require('./RateLimiterStoreAbstract');
+const RateLimiterRes = require('./RateLimiterRes');
+
+const MAX_TRANSACTION_TRIES = 5;
+
+class RateLimiterEtcd extends RateLimiterStoreAbstract {
+  /**
+   * @param {Object} opts
+   */
+  constructor(opts) {
+    super(opts);
+
+    if (!opts.storeClient) {
+      throw new Error('You need to set the option "storeClient" to an instance of class "Etcd3".');
+    }
+
+    this.client = opts.storeClient;
+  }
+
+  /**
+   * Get RateLimiterRes object filled depending on storeResult, which specific for exact store.
+   */
+  _getRateLimiterRes(rlKey, changedPoints, result) {
+    const res = new RateLimiterRes();
+
+    res.isFirstInDuration = changedPoints === result.points;
+    res.consumedPoints = res.isFirstInDuration ? changedPoints : result.points;
+    res.remainingPoints = Math.max(this.points - res.consumedPoints, 0);
+    res.msBeforeNext = result.expire ? Math.max(result.expire - Date.now(), 0) : -1;
+
+    return res;
+  }
+
+  /**
+   * Resolve with object used for {@link _getRateLimiterRes} to generate {@link RateLimiterRes}.
+   */
+  async _upsert(rlKey, points, msDuration, forceExpire = false) {
+    const expire = msDuration > 0 ? Date.now() + msDuration : null;
+
+    let newValue = { points, expire };
+    let oldValue;
+
+    // If we need to force the expiration, just set the key.
+    if (forceExpire) {
+      return this.client
+        .put(rlKey)
+        .value(JSON.stringify(newValue))
+        .then(newValue);
+    }
+
+    // Try to add a new key (returns false if the key was already there)
+    const added = await this.client
+      .if(rlKey, 'Version', '===', '0')
+      .then(this.client
+        .put(rlKey)
+        .value(JSON.stringify(newValue)))
+      .commit()
+      .then(result => !!result.succeeded);
+
+    // If the key already exists, try to update it in a transaction
+    if (!added) {
+      let success = false;
+
+      for (let i = 0; i < MAX_TRANSACTION_TRIES; i++) {
+        // eslint-disable-next-line no-await-in-loop
+        oldValue = await this._get(rlKey);
```

Comment:
@Tobias4872 Could you clarify how it is working in transaction? As I understand it gets data and then puts it. There is some time between get and put when the second get request may be sent to Ectd and it will return the same oldValue as the first one. Is that correct?

Reply:
It first pulls the old value and then sets the new value conditionally - it will only set the new value in case the old value is still present. So, let's assume the key holds the value 1. It will pull that value first and hold it in `oldValue`. And now the magic happens. When putting the new value (let's assume 2), it will first check if the old value 1 is still present. If it is it sets the value 2 and the transaction will succeed. If the value in the store is e.g. 3, the comparison will fail, and the logic starts again up to MAX_TRANSACTION_TRIES. Hope this makes sense - unfortunately, there is no atomic increase or add command in etcd.

Reply:
It is documented [here](https://etcd.io/docs/v3.4/learning/api/).

Reply:
Interesting. Thank you for the explanation.
It is going to be some kind of bottleneck under the high traffic since only one transaction can succeed during a period of time.

Question: Do you need atomic increments? If not you could rename the class to RateLimiterEtcdNonAtomic and implement simple get/set logic. Let me know what you think about that. We could have both `RateLimiterEtcd` and `RateLimiterEtcdNonAtomic` if that makes sense.

Reply:
Good idea :-) I have implemented that now by adding the suggested `RateLimiterEtcdNonAtomic` and by using that is the base class for `RateLimiterEtcd` which just overrides the `_upsert` method with the "transactional" logic.

Reply:
Good idea!